src/main/java/org/elasticsearch/index/query/BaseQueryBuilder.java

```diff
@@ -68,5 +69,7 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         return builder;
     }
 
+    public abstract Query toQuery(QueryParseContext parseContext) throws QueryParsingException, IOException;
```

Comment:
I see that where we implement this method we either do e.g. `return new BoostingQueryParser().parse(parseContext);` for leaf queries or e.g. `return parseContext.indexQueryParserService().queryParser(BoolQueryParser.NAME).parse(parseContext);` for compound queries. 

Can we do the same thing in both cases? I see two options:

1) get it from the indexQueryParserService all the time given the name of the parser then, for leaf queries too? We could have a default impl for the method in that case and the only connection required between builder and parser would become the parser name.

2) create new instances (not sure if possible in all cases) manually instead of relying on indexQueryParserService

Thoughts?